Implementation necessary to achieve use case 3 - obtaining events

alien-mcl opened the pull request:
Please find this PR approaching use case 3 - obtaining events.
While the changes in the implementation are relatively small, this PR should open a discussion on how the public API should look like.
I personally prefer a bag of hypermedia controls from which developer can filter out what he needs. Alternative is to have specialized methods that will pick i.e. links or operations.
Drawback of the latter from my perspective is that we may find nuances that link may be considered a special kind of operation or other interesting observations.
Also suggested implementation uses the client instance directly to obtain resources - alternative is to have hypermedia controls to expose methods for doing so.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/hydracg/heracles.ts/11)
<!-- Reviewable:end -->

alien-mcl commented:
It seems this PR went unnoticed. Is there any criticism about changes provided?

lanthaler commented:
I was on vacation, too bad no one else picked it up in the meantime. Feel free to ping the mailing list for each new PR. It might be that most people still aren't subscribed to the repo and thus won't get a notification.

---

Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 37 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4L7rxiY_q07zeFLqw:-Kq4L7rxiY_q07zeFLqx:b1qr1gu) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L37)):*
> ```XML
>                 beforeEach(run(async function () {
>                     this.events = await this.client.getResource(this.entryPoint.hypermedia
>                         .find(link => ~link.isA.indexOf(hydra.Collection)).iri);
> ```

What is this tilde (`~`) for here?

---

*[integration-tests/HydraClient.spec.ts, line 38 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LHdqhIF3yVV4fWux:-Kq4LHdqhIF3yVV4fWuy:b-cp6i5b) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L38)):*
> ```XML
>                     this.events = await this.client.getResource(this.entryPoint.hypermedia
>                         .find(link => ~link.isA.indexOf(hydra.Collection)).iri);
>                     this.members = this.events.hypermedia.find(link => ~link.isA.indexOf(hydra.Collection)).members;
> ```

This interface seems very un-intuitive to me. As a user of Heracles I'd expect to be able to do something like `this.events.get(hydra.member)` or even `this.events.getMembers()`

---

*[integration-tests/HydraClient.spec.ts, line 42 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LnoKi2BcOODhZI_o:-Kq4LnoKi2BcOODhZI_p:b46wmg8) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L42)):*
> ```XML
> 
>                 it("should obtain a collection of events", function () {
>                     expect(this.members).toEqual([{ iri: this.url + "api/events/1" }]);
> ```

Given the description, we should probably type the members and check whether there are a few events instead of looking for a specific URL.

---

*[integration-tests/server/api/events.jsonld, line 6 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4M1_4KXKvyIzj3m4t:-Kq4M1_4KXKvyIzj3m4u:bda3qif) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/server/api/events.jsonld#L6)):*
> ```JSONLD
>     "@type": "hydra:Collection",
>     "member": [
>         { "@id": "/api/events/1" }
> ```

As already mentioned in the test, I think we should type this and perhaps add 1 or 2 more members.

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 140 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4NNh3tGjctqSWL1IC:-Kq4NNh3tGjctqSWL1ID:b-gkq3xm) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L140)):*
> ```XML
>         for (let resource of result)
>         {
>             JsonLdHypermediaProcessor.fixTypeOf(resource);
> ```

Did you forget to remove the method above or the code below? You are doing the same thing twice.

I would also recommend to rename this method to something like makeTypeArray or something similar that is a bit more descriptive.

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 34 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4O_GVApj-bqBckEU-:-Kq4O_GVApj-bqBckEU0:b-7gthpn) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L34)):*
> ```XML
>             }));
> 
>             it("should process data", function() {
> ```

Is this nesting off? You seem to close the describe block above

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4OzXeYshR9U1VHduK:-Kq4OzXeYshR9U1VHduL:b723e18) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L39)):*
> ```XML
> 
>             it("should separate hypermedia", function() {
>                 expect(this.result.hypermedia).toEqual([
> ```

I think this shows the leaky nature of this interface quite clearly. It returns a more or less random set of data. I'd expect to be able to get a list of links to related resources and a list of operations that I then can invoke.

Something like

```
links = this.result.getLinks();
--> returns an array of Link instances which have the property, the target 
    (and perhaps a few additional properties describing the target)
links[0].getProperty() --> examplevocab:events
links[0].getTarget() --> /api/events
...
```

This link could then simply be passed to the client `this.client.getResource(links[0])` or, perhaps more intuitive, `links[0].retrieveTarget(this.client)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-:-Kq4JB0Ej5lHisMHaONb:bea3x02)*
<!-- Sent from Reviewable.io -->

alien-mcl commented:
No problem - I expected it somehow

---

Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 37 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4L7rxiY_q07zeFLqw:-KqZfI-G9uwNWj2flxWw:b-qpma3k) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L37)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

What is this tilde (`~`) for here?
</blockquote></details>

Nicely converts result of the indexOf [-1; n] method into a boolean

---

*[integration-tests/HydraClient.spec.ts, line 38 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LHdqhIF3yVV4fWux:-KqZfTGbs_ds_FBRFO5D:btnjbgr) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L38)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

This interface seems very un-intuitive to me. As a user of Heracles I'd expect to be able to do something like `this.events.get(hydra.member)` or even `this.events.getMembers()`
</blockquote></details>

It was the cheapest. As for your example - event is the original resource - we agreed that by default we don't modify it. I'd put it to the hypermedia property. As for the members, I think get(hydra.member) seems more generic as hypermedia is a bag of various objects.

---

*[integration-tests/HydraClient.spec.ts, line 42 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LnoKi2BcOODhZI_o:-KqZfsLC3uURtNemL1RI:b-orumjy) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L42)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Given the description, we should probably type the members and check whether there are a few events instead of looking for a specific URL.
</blockquote></details>

Good point

---

*[integration-tests/server/api/events.jsonld, line 6 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4M1_4KXKvyIzj3m4t:-KqZg1zsba9ci6NIUj11:bsw378a) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/server/api/events.jsonld#L6)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

As already mentioned in the test, I think we should type this and perhaps add 1 or 2 more members.
</blockquote></details>

You mean have it inlined in assertions?

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 140 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4NNh3tGjctqSWL1IC:-KqZgOb4p8EMRfsEgtfh:bl5qml7) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L140)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Did you forget to remove the method above or the code below? You are doing the same thing twice.

I would also recommend to rename this method to something like makeTypeArray or something similar that is a bit more descriptive.
</blockquote></details>

Indeed, something's redundant here

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 34 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4O_GVApj-bqBckEU-:-KqZgdosGnOzUTwDFT9Q:b-agm9vw) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L34)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Is this nesting off? You seem to close the describe block above
</blockquote></details>

I'm not sure I get it

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4OzXeYshR9U1VHduK:-KqZgwB3JNOfU-kQSrV5:bm3isv2) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L39)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

I think this shows the leaky nature of this interface quite clearly. It returns a more or less random set of data. I'd expect to be able to get a list of links to related resources and a list of operations that I then can invoke.

Something like

```
links = this.result.getLinks();
--> returns an array of Link instances which have the property, the target 
    (and perhaps a few additional properties describing the target)
links[0].getProperty() --> examplevocab:events
links[0].getTarget() --> /api/events
...
```

This link could then simply be passed to the client `this.client.getResource(links[0])` or, perhaps more intuitive, `links[0].retrieveTarget(this.client)`

</blockquote></details>

I'll think it over

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-:-KqZfyPO-lqvAsQHeY-E:b-5pwl2u)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 37 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4L7rxiY_q07zeFLqw:-Kr7OZ1S2Jl5QxRZklIJ:b-j32f71) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L37)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Nicely converts result of the indexOf [-1; n] method into a boolean
</blockquote></details>

Could we please make this more explicit? The tilde is a bitwise NOT operator so -1 becomes 0, 0 becomes -1, 1 becomes -2.. not easy to follow

---

*[integration-tests/HydraClient.spec.ts, line 38 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LHdqhIF3yVV4fWux:-Kr7PVCC8wLMpyCG6md_:bg4fx7z) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L38)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

It was the cheapest. As for your example - event is the original resource - we agreed that by default we don't modify it. I'd put it to the hypermedia property. As for the members, I think get(hydra.member) seems more generic as hypermedia is a bag of various objects.
</blockquote></details>

Sure, it's more generic but is that a problem? I understand that it is more difficult to implement but it makes the use of this library much easier.. I'm fine to leave that to a separate PR though if you prefer

---

*[integration-tests/server/api/events.jsonld, line 6 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4M1_4KXKvyIzj3m4t:-Kr7PzDKlrRvPKlC14lT:b-u6tdgn) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/server/api/events.jsonld#L6)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

You mean have it inlined in assertions?
</blockquote></details>

I meant changing it to something like

```
"member": [
        { "@id": "/api/events/1", "@type": "Event" },
        { "@id": "/api/events/2", "@type": "Event" },
        { "@id": "/api/events/3", "@type": "Event" }
```

---

*[src/DataModel/JsonLd/context.json, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kr7RqC3CMrHeIOAVJoT:-Kr7RqC4ibFeig3T6Eud:bjy06vu) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/context.json#L26)):*
> ```JSON
>         },
>         "http://www.w3.org/ns/hydra/core#Collection": "http://www.w3.org/ns/hydra/core#Collection",
>         "http://www.w3.org/ns/hydra/core#Resource": "http://www.w3.org/ns/hydra/core#Resource"
> ```

I assume this should be

```
        "Collection": "hydra:Collection",
        "Resource": "hydra:Resource"
```

instead

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 34 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4O_GVApj-bqBckEU-:-Kr7QUHRpMKMzqh_7IK1:bmt044t) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L34)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

I'm not sure I get it
</blockquote></details>

Sorry, my bad.. I missed the `run` and inside the `it` block. I thought you close the `describe` block.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

elf-pavlik commented:
*[integration-tests/HydraClient.spec.ts, line 37 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4L7rxiY_q07zeFLqw:-KrnAv76Lf2I7D5vUKzO:bsa359y) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L37)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Could we please make this more explicit? The tilde is a bitwise NOT operator so -1 becomes 0, 0 becomes -1, 1 becomes -2.. not easy to follow
</blockquote></details>

stepping aside from the code style aspect, it looks like client just picks any first collection that appears in the entry point with `.find(link => ~link.isA.indexOf(hydra.Collection)).iri)`, i think test should include at least one other collection (eg. '/api/people') so that client needs to discover the desired collection
if this PR doesn't want to address discovery, maybe it should just hard code `/api/events` just as "should obtain a collection of events" does

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

alien-mcl commented:
Review status: 2 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 37 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4L7rxiY_q07zeFLqw:-Ktmdk4PdIG09ZXrzEDd:b-taizwq) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L37)):*
<details><summary><i>Previously, elf-pavlik (elf Pavlik) wrote…</i></summary><blockquote>

stepping aside from the code style aspect, it looks like client just picks any first collection that appears in the entry point with `.find(link => ~link.isA.indexOf(hydra.Collection)).iri)`, i think test should include at least one other collection (eg. '/api/people') so that client needs to discover the desired collection
if this PR doesn't want to address discovery, maybe it should just hard code `/api/events` just as "should obtain a collection of events" does
</blockquote></details>

Ok, I've removed this pseudo-discovery and replaced it with explicit resource Url.

---

*[integration-tests/HydraClient.spec.ts, line 38 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LHdqhIF3yVV4fWux:-Ktmdr2X3geD9MQXEIn3:b8f54q1) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L38)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Sure, it's more generic but is that a problem? I understand that it is more difficult to implement but it makes the use of this library much easier.. I'm fine to leave that to a separate PR though if you prefer
</blockquote></details>

I've added `resource.hypermedia.members` with option to expand with other accessors.

---

*[integration-tests/HydraClient.spec.ts, line 42 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LnoKi2BcOODhZI_o:-Ktmdxg2SI8ZwQ7u3G4J:baytayv) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L42)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Good point
</blockquote></details>

Replaced with typed member check

---

*[integration-tests/server/api/events.jsonld, line 6 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4M1_4KXKvyIzj3m4t:-KtmeF-0ER275M7Qo1GL:b-kcinmo) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/server/api/events.jsonld#L6)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

I meant changing it to something like

```
"member": [
        { "@id": "/api/events/1", "@type": "Event" },
        { "@id": "/api/events/2", "@type": "Event" },
        { "@id": "/api/events/3", "@type": "Event" }
```
</blockquote></details>

I've added types and two more members

---

*[src/DataModel/JsonLd/context.json, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kr7RqC3CMrHeIOAVJoT:-KtmeU4Ys1MX0Grjq85b:b-yku4ga) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/context.json#L26)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

I assume this should be

```
        "Collection": "hydra:Collection",
        "Resource": "hydra:Resource"
```

instead
</blockquote></details>

Ahh. It was some time ago, but I think this was to force either framing or flattening to enforce those urls. For some reason these were incorrectly processed.

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 34 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4O_GVApj-bqBckEU-:-KtmehKQCuqJQSvDvjVp:b32w1t) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L34)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Sorry, my bad.. I missed the `run` and inside the `it` block. I thought you close the `describe` block.
</blockquote></details>

Ok

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Reviewed 9 of 11 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 38 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LHdqhIF3yVV4fWux:-KtwmEGwtnFZnPs7TLaC:b-zdzh23) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L38)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

I've added `resource.hypermedia.members` with option to expand with other accessors.
</blockquote></details>

OK. I'm still not very happy about this `hypermedia` property but let's keep it as is for now and move on.

---

*[integration-tests/server/api/events.jsonld, line 6 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4M1_4KXKvyIzj3m4t:-KtwlUTMaVluBRlTiqTR:b-zcjvte) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/server/api/events.jsonld#L6)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

I've added types and two more members
</blockquote></details>

Thanks

---

*[src/DataModel/JsonLd/context.json, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kr7RqC3CMrHeIOAVJoT:-KtwlpJGtwRDiIgM8eWw:byz07q0) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/context.json#L26)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Ahh. It was some time ago, but I think this was to force either framing or flattening to enforce those urls. For some reason these were incorrectly processed.
</blockquote></details>

I don't follow. What's the point of mapping these URLs to themselves?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

elf-pavlik commented:
*[src/DataModel/JsonLd/context.json, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kr7RqC3CMrHeIOAVJoT:-Ku04cRngh6GorO0hmA8:b-3xb91g) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/context.json#L26)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

I don't follow. What's the point of mapping these URLs to themselves?
</blockquote></details>

It looks that in rest of the codebase you always `import { hydra } from './namespaces'` so JSON-LD context used by the client doesn't need to create any aliases. How about just removing those last two line all together?
A future PR could propose any further improvements on what uses `namespaces` module and what relies on aliases in JSON-LD context.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

elf-pavlik commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 140 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4NNh3tGjctqSWL1IC:-Ku06NuFcynUXcetCAZ0:bda9q6d) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L140)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Indeed, something's redundant here
</blockquote></details>

I think you might have extracted `fixTypeOf` from `fixType` but didn't delete extracted code after calling `fixTypeOf` inside of `fixType`.
BTW in cases other than `@type` JSON-LD context seems to rely on `"@container": "@set"`to enforce array. It does not seem possible to do something similar for `@type`
```
    "isA": {
      "@id": "@type",
      "@container": "@set"
    }
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

alien-mcl commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 38 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4LHdqhIF3yVV4fWux:-Ku0SalOoWgYLtCzaLbq:b32w0x) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/integration-tests/HydraClient.spec.ts#L38)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

OK. I'm still not very happy about this `hypermedia` property but let's keep it as is for now and move on.
</blockquote></details>

OK

---

*[src/DataModel/JsonLd/context.json, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kr7RqC3CMrHeIOAVJoT:-Ku0S9LBc_6NT8iNTwaT:brn556l) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/context.json#L26)):*
<details><summary><i>Previously, elf-pavlik (elf Pavlik) wrote…</i></summary><blockquote>

It looks that in rest of the codebase you always `import { hydra } from './namespaces'` so JSON-LD context used by the client doesn't need to create any aliases. How about just removing those last two line all together?
A future PR could propose any further improvements on what uses `namespaces` module and what relies on aliases in JSON-LD context.
</blockquote></details>

As for those two full urls - I'm using a hydra's Url as `@vocab` which causes hydra based types to be shortened to `Collection` or `Resource`. Still, while processing full urls are preferred, thus that's why I'm forcing those in the context.
As for the namespace imports - I'm not sure what does it have to do with JSON-LD processing - imports are just code imports.

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 140 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4NNh3tGjctqSWL1IC:-Ku0RTmRFFLN7-mxfQZF:bpguyht) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L140)):*
<details><summary><i>Previously, elf-pavlik (elf Pavlik) wrote…</i></summary><blockquote>

I think you might have extracted `fixTypeOf` from `fixType` but didn't delete extracted code after calling `fixTypeOf` inside of `fixType`.
BTW in cases other than `@type` JSON-LD context seems to rely on `"@container": "@set"`to enforce array. It does not seem possible to do something similar for `@type`
```
    "isA": {
      "@id": "@type",
      "@container": "@set"
    }
```
</blockquote></details>

Yeah - I've tried using @set for @type. I've raised it at https://github.com/digitalbazaar/jsonld.js/issues/191 (unsure whether it is implementation or specification issue). BTW - I forgot to fix - I'll try to update PR ASAP.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

elf-pavlik commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 12 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/DataModel/JsonLd/context.json, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kr7RqC3CMrHeIOAVJoT:-Ku0qYtdKvJxPUut8_2J:bfmill8) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/context.json#L26)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

As for those two full urls - I'm using a hydra's Url as `@vocab` which causes hydra based types to be shortened to `Collection` or `Resource`. Still, while processing full urls are preferred, thus that's why I'm forcing those in the context.
As for the namespace imports - I'm not sure what does it have to do with JSON-LD processing - imports are just code imports.
</blockquote></details>

I think we shouldn't let this block this PR so I captured it in separate issue https://github.com/HydraCG/Heracles.ts/issues/15

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 140 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4NNh3tGjctqSWL1IC:-Ku0rh5PTKH9t3bcw4Y6:brq0jlr) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L140)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Yeah - I've tried using @set for @type. I've raised it at https://github.com/digitalbazaar/jsonld.js/issues/191 (unsure whether it is implementation or specification issue). BTW - I forgot to fix - I'll try to update PR ASAP.
</blockquote></details>

regarding changing naming of this function, maybe we could propose fine tuning it after merging this PR with a dedicated PR?

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4OzXeYshR9U1VHduK:-Ku0qrKPupOk8GCuBsbm:b1sg6tn) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L39)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

I'll think it over
</blockquote></details>

can we possibly capture it into a separate issue so that we can merge this PR and keep iterating?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-:-Ku0sClOHQloxHprtyBN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 140 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4NNh3tGjctqSWL1IC:-KuG-VngDAuaq8ored-h:b70lxvd) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L140)):*
<details><summary><i>Previously, elf-pavlik (elf Pavlik) wrote…</i></summary><blockquote>

regarding changing naming of this function, maybe we could propose fine tuning it after merging this PR with a dedicated PR?
</blockquote></details>

Using `@set` for `@type` doesn't work as JSON-LD doesn't allow you to change the behavior of its keywords.

Tweaking the naming in a dedicated PR sounds good to me.

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-Kq4OzXeYshR9U1VHduK:-KuG-suZna7K_aYtDwIU:bw3awgo) ([raw file](https://github.com/hydracg/heracles.ts/blob/70d74b9a8ecfa5504c12711e3fc9fd48244cf4f4/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L39)):*
<details><summary><i>Previously, elf-pavlik (elf Pavlik) wrote…</i></summary><blockquote>

can we possibly capture it into a separate issue so that we can merge this PR and keep iterating?
</blockquote></details>

Sure. Filed issue #16 for this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-:-KuG0sOgqNcOI_FTVzq4:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tpluscode commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/11#-:-KuL5IHeJjbYKpIZpCw3:bnfp4nl)*
<!-- Sent from Reviewable.io -->
